[SR] Repository list and details UI

jen-huang opened the pull request:
Adds snapshot repositories list view and details panel. Action buttons (create, delete, edit) in screenshots are just placeholders for now 😄 

There is no Redux usage here! React hooks have been very sufficient so far.

Details for a single repository is deep-linked via a path parameter rather than query parameter like in our other apps. This simplifies a lot of code.

<details>
<summary>Instructions for setting up basic repositories for testing</summary>

#### File system
1. Add the file system path you want to use to elasticsearch.yml or as part of starting up ES:
```
path:
  repo: /tmp/es-backups
```
or
```
yarn es snapshot --license trial -E path.repo=/tmp/es-backups
```
2. Use Console to add repository, use the file system path above as `location` setting:
```
PUT /_snapshot/my_backup
{
  "type": "fs",
  "settings": {
    "location": "/tmp/es-backups",
    "chunk_size": "10mb"
  }
}
```
3. Adjust `settings` as necessary, all available settings can be found in docs:
https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_shared_file_system_repository

#### Readonly
Readonly repositories only take `url` setting. Documentation: https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_read_only_url_repository

It's easy to set up a `file:` url:
```
PUT _snapshot/my_readonly_repository
{
  "type": "url",
  "settings": {
    "url": "file:///tmp/es-backups"
  }
}
```

### Source only
Source only repositories are special in that they are basically a wrapper around another repository type. Documentation: https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_source_only_repository

This means that the settings that are available depends on the `delegate_type` parameter. For example, this source only repository delegates to `fs` (file system) type, so all file system rules and available settings apply:
```
PUT _snapshot/my_src_only_repository
{
  "type" : "source",
  "settings" : {
    "delegate_type" : "fs",
    "location" : "/tmp/es-backups"
  }
}
```

#### Plugin-based repositories:
There are four official repository plugins available: S3, GCS, HDFS, Azure. Please refer to the meta ticket **Developer notes** section : https://github.com/elastic/elasticsearch-team/issues/118

Available plugin repository settings can be found in docs: https://www.elastic.co/guide/en/elasticsearch/plugins/master/repository.html
</details>

No repositories prompt:
![image](https://user-images.githubusercontent.com/1965714/54467144-941e3f00-4740-11e9-8a96-13d26ad01c21.png)

List of repositories, with ability to filter by type:
![image](https://user-images.githubusercontent.com/1965714/54467179-c62fa100-4740-11e9-900a-003c1680467b.png)

Repository detail panel:
![image](https://user-images.githubusercontent.com/1965714/54467195-e5c6c980-4740-11e9-890d-653c96761b46.png)

Loading repositories:
![image](https://user-images.githubusercontent.com/1965714/54467374-ef9cfc80-4741-11e9-8be2-a51d92017689.png)

Repositories error:
![image](https://user-images.githubusercontent.com/1965714/54467406-2246f500-4742-11e9-9c51-52cd3e8d003f.png)

Repository detail panel loading:
![image](https://user-images.githubusercontent.com/1965714/54467488-9d101000-4742-11e9-8b80-e831b02de9bf.png)

Repository detail panel error:
![image](https://user-images.githubusercontent.com/1965714/54467454-66d29080-4742-11e9-9cba-a9c4006fc939.png)

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5699/)

jen-huang commented on x-pack/plugins/infra/types/eui.d.ts:
```diff
@@ -116,6 +116,8 @@ declare module '@elastic/eui' {
     loading?: any;
     hasActions?: any;
     message?: any;
+    rowProps?: any;
```
VSCode was complaining about missing props on `EuiInMemoryTable`, it used the definition from here

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5835/)

jen-huang commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5850/)

cjcenizal commented on x-pack/plugins/snapshot_restore/public/app/services/api.ts:
```diff
@@ -0,0 +1,68 @@
+      if (response.data) {
+        setData(response.data);
+      } else {
+        setError(new Error(response.statusText));
```
Minor nit: you could `throw` the error and reverse the order of the conditions to simplify things a bit.

```js
      if (!response.data) {
        throw new Error(response.statusText);
      }
      setData(response.data);
```

cjcenizal commented on x-pack/plugins/snapshot_restore/public/app/services/api.ts:
```diff
@@ -0,0 +1,68 @@
+  const [loading, setLoading] = useState<boolean>(false);
+  const [data, setData] = useState<any>([]);
+
+  const fetch = async () => {
```
The consumer provides `method` to this hook so it can also be used to send `POST` etc requests, and "fetch" seems more geared towards retrieving data (it's also already part of the [global namespace](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)). How about renaming this to `request` and `useRequest` instead?

cjcenizal commented on x-pack/plugins/snapshot_restore/public/app/components/repository_type_name.tsx:
```diff
@@ -0,0 +1,113 @@
+  ] = useAppState() as [AppStateInterface];
+
+  const getTypeName = (repositoryType: RepositoryType): JSX.Element => {
+    switch (repositoryType) {
```
Another minor nit, you could use a map instead of a switch to reduce visual noise a bit:

```js
    const map = {
      [FSRepositoryType]: (
        <FormattedMessage
          id="xpack.snapshotRestore.repositoryType.fileSystemTypeName"
          defaultMessage="Shared File System"
        />
      ),

      [ReadonlyRepositoryType]: (
        <FormattedMessage
          id="xpack.snapshotRestore.repositoryType.readonlyTypeName"
          defaultMessage="Read-only URL"
        />
      ),

      [S3RepositoryType]: (
        <FormattedMessage
          id="xpack.snapshotRestore.repositoryType.s3TypeName"
          defaultMessage="AWS S3"
        />
      ),

      /* ... */      
    };
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6059/)

jen-huang commented:
Thanks for the feedback @cjcenizal! I've applied all your suggestions and added a collapsed section to the PR description about setting up repositories. Hope that helps.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6068/)

jen-huang commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6077/)

sebelga commented on x-pack/plugins/snapshot_restore/public/app/components/repository_delete_provider.tsx:
```diff
@@ -0,0 +1,27 @@
+  if (isModalOpen && repositoryNames.length) {
+    /* placeholder */
+  }
+
```
This is how I would have typed the children props:

```ts
interface Props {
  children: (fun: deleteRepositoryFunction) => ReactNode;
}

type deleteRepositoryFunction = (names: Array<Repository['name']>) => void;

export const RepositoryDeleteProvider = ({ children }: Props) => {
  const [repositoryNames, setRepositoryNames] = useState<Array<Repository['name']>>([]);
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const deleteRepository: deleteRepositoryFunction = names => {
    setIsModalOpen(true);
    setRepositoryNames(names);
  };

  ...

```

jen-huang commented:
Thanks for the feedback and great discussion around TS, @sebelga! I've updated the PR with most of your suggestions, except I added another `enum` for types to be used for conditionals instead of string literals. As we discussed, I'm deferring the folder structure for a later team consensus as we gain more knowledge around using React contexts and hooks.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6266/)

jen-huang commented on x-pack/plugins/snapshot_restore/common/repository_types.ts:
```diff
@@ -0,0 +1,137 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export const FSRepositoryType = 'fs';
```
I think you are viewing an outdated file 😄 I moved this to `common/types/` and updated with enums: https://github.com/elastic/kibana/pull/33367/files#diff-5c7d6779b8a84817b082813365634361R1

sebelga replied:
Yeah sorry! I wrote the comment _before_ you merged it. Then left the review for a while and we got a race condition! 😄

sebelga commented:
@jen-huang I created a separate PR to discuss some change proposal. Let's discuss them over there 😊 https://github.com/elastic/kibana/pull/33690
